default/python/universe_generation/galaxy.py

```diff
@@ -1,35 +1,297 @@
 import sys
-from random import random, uniform, randint, gauss, choice
-from math import pi, sin, cos, sqrt
+from random import random, uniform, randint, gauss
+from math import pi, sin, cos, acos, sqrt, ceil, floor
+from collections import defaultdict
 
 import freeorion as fo
 import util
 import universe_tables
 
 
-class AdjacencyGrid:
+class AdjacencyGrid(object):
     def __init__(self, universe_width):
+        assert universe_tables.MIN_SYSTEM_SEPARATION < universe_tables.MAX_STARLANE_LENGTH
         self.min_dist = universe_tables.MIN_SYSTEM_SEPARATION
-        self.cell_size = max(universe_width / 50, self.min_dist)
+        self.max_dist = universe_tables.MAX_STARLANE_LENGTH
+        self.cell_size = min(max(universe_width / 50, self.min_dist), self.max_dist/sqrt(2))
         self.width = int(universe_width / self.cell_size) + 1
-        self.grid = [[[] for _ in range(self.width)] for _ in range(self.width)]
+        self.grid = defaultdict(list)
         print "Adjacency Grid: width", self.width, ", cell size", self.cell_size
 
+    def cell(self, pos):
+        return (int(pos[0] / self.cell_size), int(pos[1] / self.cell_size))
+
     def insert_pos(self, pos):
-        self.grid[int(pos[0] / self.cell_size)][int(pos[1] / self.cell_size)].append(pos)
+        self.grid[self.cell(pos)].append(pos)
+
+    def remove(self, pos):
```

Comment:
Yep. Either call it `remove_pos` or rename `insert_pos` to `insert`. I strongly prefer the former, as it more clearly says what it does (inserting/removing a *position*).